### PR TITLE
2236 Optimize acces to document lines in incremental mode

### PR DESCRIPTION
**Pull request by smedilol**
For WI #2236.
Optimize access to our document lines (`ISearchableReadOnlyList`) in incremental mode.

`ISearchableReadOnlyList` is currently implemented with an `ImmutableList`.
`ImmutableList` is great to manage our `CodeElementLines` (see `TypeCobol.Compiler.CompilationDocument.compilationDocumentLines`) and update them in incremental mode. But the problem is `ImmutableList` are very slow to iterate. A direct element access is also very slow.
After tests, it's faster to iterate our `ImmutableList` only once and copy it into a standard List then iterate this List multiple times, than to iterate the `ImmutableList` multiple times.

This PR is divided in 3 steps:
- Replace cast to ImmutableList with cast to ISearchableReadOnlyList (0b05c937e6146842b89e0da358b94fbeabf805d5).
  - This PR don't try to replace all usage of CodeElementsLine by ICodeElementsLine
- Code cleanup in  ISearchableReadOnlyList to help following step (c8607ada6275307656f0b285fa8473fc54c41912).
- Create implementation of ISearchableReadOnlyList with a simple List and use it in `TypeCobol.Compiler.CompilationDocument.RefreshTokensDocumentSnapshot` (168a3ca77b907f0f6655c07b36bd71f992716097).

## What's left to do
- Specific incremental tests. We already have some incremental tests. I need to be sure that they cover the case of this implementation.
I've tested in our private IDE and it works fine so far.
   - Check line in CompilationDocument that used `.ToBuilder()` then `.ToImmutable()`.
   - Check `TypeCobol.Compiler.CompilationDocument.PrepareDocumentLineForUpdate` and direct access to `compilationDocumentLines[index]`

## Bonus
We could add a new parameter to know if the parser will be used in incremental mode or not and then directly use `SearchableReadOnlyList` for batch mode to handle `TypeCobol.Compiler.CompilationDocument.compilationDocumentLines`.
Or always start with `SearchableReadOnlyList` and if incremental mode is triggered, switch to `ImmutableList`.

## Performance gain
Small gain only in incremental mode.
[Perf-Gain.xlsx](https://github.com/TypeCobolTeam/TypeCobol/files/9899632/Perf-Gain.xlsx)
This Excel also include gain combined with PR #2370.

As always with performance gain, if you run the same test twice, you won't gain exactly the same results.

### CODE REVIEW

File: TypeCobol/Compiler/CompilationDocument.cs [hunk 1]
```diff
@@ -597,7 +597,9 @@ public void RefreshTokensDocumentSnapshot()
                 // Create a new snapshot only if things changed since last snapshot
                 if (TokensDocumentSnapshot == null || TokensDocumentSnapshot.CurrentVersion != currentTokensLinesVersion)
                 {
-                    TokensDocumentSnapshot = new TokensDocument(TextSourceInfo, textLinesVersionForCurrentTokensLines, currentTokensLinesVersion, compilationDocumentLines.ToImmutable());
+                    //Copy our compilationDocumentLines into a new List.
+                    //As compilationDocumentLines is an ImmutableList, it's faster in next step to copy the ImmutableList once in a new list and then iterate over it multiples times.
```
Comment (fm-117) [inline TypeCobol/Compiler/CompilationDocument.cs:601]: To be precise, an `ImmutableList.Builder`

File: TypeCobol/Compiler/Concurrency/SearchableReadOnlyList.cs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace TypeCobol.Compiler.Concurrency
+{
+
+
+    /// <summary>
+    /// TODO it's not a real ReadOnyList, try to inherits from ReadOnlyCollection or override all *modification* method and a throw an exception.
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol/Compiler/Concurrency/SearchableReadOnlyList.cs:13]: We can also wrap a List ?

File: TypeCobol/Compiler/Concurrency/SearchableReadOnlyList.cs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+                if (this._index < 0 || (uint)this._index >= (uint)list.Count)
+                    return this.MoveNextRare();
+                this._current = list[this._index];
+                this._index += _reverse ? -1 : 1;
+                return true;
+            }
+
+            private bool MoveNextRare()
+            {
+                this._index = this._list.Count + 1;
```
Comment (fm-117) [inline TypeCobol/Compiler/Concurrency/SearchableReadOnlyList.cs:137]: `+ 1` ?

**Review by fm-117 — CHANGES_REQUESTED** [review]

**Comment by smedilol** [comment]
@fm-117 @efr15 This PR must be tested again with .Net 6.
I'm not sure it's still worth it.

**Comment by smedilol** [comment]
@fm-117 @efr15 Maybe just keep 2 commits:
- Replace cast to ImmutableList with cast to ISearchableReadOnlyList (https://github.com/TypeCobolTeam/TypeCobol/commit/0b05c937e6146842b89e0da358b94fbeabf805d5).
    - This PR don't try to replace all usage of CodeElementsLine by ICodeElementsLine
- Code cleanup in ISearchableReadOnlyList to help following step (https://github.com/TypeCobolTeam/TypeCobol/commit/c8607ada6275307656f0b285fa8473fc54c41912).

**Comment by fm-117** [comment]
Replaced by #2575.

**Pull request closed without merging**